### DISCUSSION TITLE
Does XRDP support printer redirection now?

### DISCUSSION THREAD

**Discussion by AlekseySigachev** [neutrinolabs/xrdp#2982] [Q&A] [answered]
Can you please tell me the version whitch I must use to achive printer redirection?

**Comment by matt335672** [comment, marked as answer]
We don't support it I'm afraid.

**Comment by lakshmanpasala** [comment]
I am working on a patch for supporting printer redirect via a cups virtual printer. I will submit a PR for others if they are interested. I am adding this to the `devredir.c` file itself and a separate file for the cups backend to receive the print jobs from the virtual printer.

**Reply by wqlxx** [reply]
> I am working on a patch for supporting printer redirect via a cups virtual printer. I will submit a PR for others if they are interested. I am adding this to the `devredir.c` file itself and a separate file for the cups backend to receive the print jobs from the virtual printer.

I can't wait to see your PR!!!!!!!!

**Reply by lakshmanpasala** [reply]
Finally got it running. Will test across different clients and printers and then submit a PR.

**Reply by lakshmanpasala** [reply]
Update is that RAW printing works but working on support XPS format.

**Reply by lakshmanpasala** [reply]
Got XPS working by converting different formats -> PDF -> XPS 

Not sure this is the most efficient way as a bunch of extra ghostscript packages had to be installed. Have tested with xrdp on combinations of ubuntu 20.04, 22.04 and 24.04 and clients being remmina and mstsc, and both XPS and RAW format print jobs work with external printers as well as save to PDF/XPS.

**Reply by lakshmanpasala** [reply]
I will create an initial draft PR for this later this month.

**Reply by matt335672** [reply]
Thanks @lakshmanpasala - looking forward to it.

**Reply by kadrim** [reply]
@lakshmanpasala can you publish your work somewhere? did not find it any of your repos.

i am in desperate need for this feature :-)

**Comment by kadrim** [comment]
nevermind, i used the weekend to create a patched xrdp version (specifically for ubuntu 24.04) that follows a comparable path as the one from @lakshmanpasala 

in the default settings, each xrdp-session spins up an isolated cupsd and adds all published printers. XPS is also achieved by intermediate PDF conversion.

you can find the repository for the patched version (alongside a prebuilt .deb) here:

https://github.com/sodgeit/xrdp-printer-redirection

as soon as i find the time, i will try to polish the code and prepare a PR for xrdp upstream

**Comment by jsorg71** [comment]
Wow very nice.
